### PR TITLE
DOC: document site.cfg.example

### LINKED ISSUE #14286: DOC, BUILD: improve documentation around site.cfg
Documentation of `site.cfg` exists in `numpy/distutils/system_info.py` but that is not extracted as a docstring into the sphinx-built documentation.

`site.cfg` is also mentioned in  `./INSTALLATION.rst.txt`, but that file is also not linked into the sphinx-built documentation, the information there should be merged with the other bits and pieces about building from source and installation.

### PR DESCRIPTION
**Pull request by mattip**
Fixes gh-14286

Add references to `site.cfg.example` to the docs, and fix the example file's `DEFAULT` section name.

### CODE REVIEW

File: doc/source/reference/distutils.rst [hunk 1]
```diff
@@ -15,6 +15,9 @@ keyword arguments to pass to the setup function (by passing the
 dictionary obtained from the todict() method of the class). More
 information is available in the :ref:`distutils-user-guide`.
 
+For configuring the preference for and location of libraries like BLAS and
```
Comment (charris) [inline doc/source/reference/distutils.rst]: Maybe something like
```
The choice and location of linked libraries such as BLAS and LAPACK as well as include paths and other such build options can be specified in a ``site.cfg`` file located in the NumPy root repository or a ``.numpy-site.cfg`` file in your home directory. See the ``site.cfg.example`` example file included in the NumPy repository or sdist for documentation.
```

File: doc/source/user/building.rst [hunk 2]
```diff
@@ -35,6 +35,9 @@ Building NumPy requires the following software installed:
    installed. However, if these are available, NumPy's setup script can detect
    them and use them for building. A number of different LAPACK library setups
    can be used, including optimized LAPACK libraries such as OpenBLAS or MKL.
+   A ``site.cfg`` file, if found in the base directory of the sources, can be
+   used to specify locations of the libraries and headers, and the preferences
+   between them. See the sample ``site.cfg.example`` for more information.
```
Comment (charris) [inline doc/source/user/building.rst]: "Between" is the wrong word here. See above for a general style suggestion. I think you can omit the "... and the... then" phrase,

Comment (mattip) [reply doc/source/user/building.rst]: fixed

File: doc/source/reference/distutils.rst (hunk 1 above)
Comment (mattip) [reply doc/source/reference/distutils.rst]: adopted

File: site.cfg.example [hunk 3]
```diff
@@ -83,7 +83,7 @@
 # This is a good place to add general library and include directories like
 # /usr/local/{lib,include}
 #
-#[ALL]
+#[DEFAULT]
```
Comment (charris) [inline site.cfg.example]: This doesn't agree with `numpy/distutils/system_info.py`, which uses `ALL`. Why the change?

Comment (charris) [reply site.cfg.example]: Seems this might break people currently using a `site.cfg` file.

**Comment by charris** [comment]
I wonder if we shouldn't use a `*.yml` file at some point?

File: site.cfg.example (hunk 3 above)
Comment (mattip) [reply site.cfg.example]: If this is wrong, why are no tests failing? I think the documentation is wrong here: the `ALL` section does **not** override other section's missing values. I reverted the section name change in 8f94b292f, and added a test that shows the documentation is wrong. Do we want to fix the documentation or add code to override missing defaults?

Comment (charris) [reply site.cfg.example]: My only concern here is back compatibility. If the code can be changed without causing problems, that is OK. If not, I think the documentation should reflect reality. A new section, maybe `DEFAULT`, could be added if that adds useful functionality :)

Comment (pv) [reply site.cfg.example]: It's true that `ALL` is the default value for `system_info.section`, but all the subclasses apart from the *_opt_info ones (which do no lookups themselves) override it. So it has no effect.

The `DEFAULT` section OTOH works, because that's builtin ConfigParser functionality.

I'm not sure if the ALL stuff has worked at some point, but currently it does not, and probably the default should be changed to `section = None` in `class system_info` to avoid confusion. Also the comment on top of `system_info.py` should be corrected.

Comment (pv) [reply site.cfg.example]: So maybe just remove this part from site.cfg.example?

Comment (mattip) [reply site.cfg.example]: It turns out the `*_opt_info` and `f2py_info` classes use `ALL`. The other classes all override the `system_info` class's `section` value with more specific names. I tried to document all this in b4ccd24

Comment (pv) [reply site.cfg.example]: The classes that use `ALL` I think don't have any tunable options.

If so, I'd suggest to not document it, and just schedule itfor removal.

Comment (pv) [reply site.cfg.example]: Sorry, this is not fully correct --- the `extra_compile_args` and `extra_link_args` in the `ALL` section may have an effect because they are added by `calc_extra_info`. The library variables don't have an effect, as `libraries` is empty list in `calc_libraries_info`.

Comment (mattip) [reply site.cfg.example]: thanks for the deep dive. I will remove ALL from the docstring.

**Comment by mkoeppe** [comment]
Another concern is that depending on a configuration file in the source directory does not work for users who want to install the package with `pip` (from a source that is not a local directory). Shouldn't a build option for `setup.py build` be added so that users can provide the location of a `site.cfg` file to be used?

**Comment by rgommers** [comment]
> I wonder if we shouldn't use a `*.yml` file at some point?

no I don't think so, this works and is fairly standard. not much point spending effort on switching

> Another concern is that depending on a configuration file in the source directory

You can use `~/.numpy-site.cfg`. Also, if you use `pip` *and* build from source with non-standard options *and* not from a local repo/dir, you're probably doing the wrong thing.

File: site.cfg.example (hunk 3 above)
Comment (dimpase) [reply site.cfg.example]: Setting [DEFAULT] libs and headers to a location different from OpenBLAS seems to break using a Debian 10 system OpenBLAS on a Python+numpy+scipy built from source in a non-standard location. Namely, if in the following `site.cfg`
```
[ALL]
library_dirs = /home/dimpase/sage-src/local/lib
include_dirs = /home/dimpase/sage-src/local/include
[blas]
include_dirs = /usr/include/x86_64-linux-gnu
blas_libs    = openblas
[lapack]
lapack_libs  = openblas
```
one adds after [ALL] section
```
[DEFAULT]
library_dirs = /home/dimpase/sage-src/local/lib
include_dirs = /home/dimpase/sage-src/local/include
```
then scipy cannot be built any more, as it cannot find blas/lapack (and it can without [DEFAULT]  section).

This is from https://trac.sagemath.org/ticket/29051 and https://trac.sagemath.org/ticket/29051

Comment (pv) [reply site.cfg.example]: That's because setting DEFAULT overrides the default search paths, which *includes* `/usr/lib` etc standard places.

File: site.cfg.example [hunk 4]
```diff
@@ -79,14 +79,12 @@
...
 # Defaults
 # ========
-# The settings given here will apply to all other sections if not overridden.
+# The settings here will apply to all sections as general defaults
 # This is a good place to add general library and include directories like
 # /usr/local/{lib,include}
-#
-#[ALL]
+#[DEFAULT]
 #library_dirs = /usr/local/lib
```
Comment (pv) [inline site.cfg.example]: Change these to `/usr/local/lib64:/usr/local/lib:/usr/lib64:/usr/lib:/usr/lib/` and `/usr/local/include:/usr/include` so it's clearer how to specify multiple paths and that default search paths will get overridden...

**Review by pv — COMMENTED** [review]
Some comments still on the DEFAULT section.

File: site.cfg.example (hunk 3 above)
Comment (dimpase) [reply site.cfg.example]: > As noted elsewhere, the search paths in ALL do not do anything at all so there's no point setting them.

OK, so that's very good to know, meaning that we may skip [ALL] and [DEFAULT] for the case at hand, at least.

File: numpy/distutils/system_info.py [hunk 5]
```diff
@@ -81,19 +30,93 @@
...
+section ALL is not intended for general use.
+
+Appropriate defaults are used if nothing is specified.
 
 The order of finding the locations of resources is the following:
  1. environment variable
  2. section in site.cfg
- 3. ALL section in site.cfg
+ 3. DEFAULT section in site.cfg
 Only the first complete match is returned.
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/distutils/system_info.py]: adding

File: site.cfg.example [hunk 6]
```diff
@@ -79,14 +79,12 @@
 
 # Defaults
 # ========
-# The settings given here will apply to all other sections if not overridden.
+# The settings here will apply to all sections as general defaults
```
[the remark this answers is not in the document]
Comment (mattip) [reply site.cfg.example]: adding

File: site.cfg.example (hunk 4 above)
Comment (mattip) [reply site.cfg.example]: adding, and adding a comment about using `os.pathsep`

**Review by pv — APPROVED** [review]

File: site.cfg.example [hunk 7]
```diff
@@ -79,14 +79,17 @@
...
 # /usr/local/{lib,include}
-#
-#[ALL]
-#library_dirs = /usr/local/lib
-#include_dirs = /usr/local/include
-#
+# These settings apply when they are not overridden in the sections below.
+# Note that the standard paths (e.g. `/usr/lib`) are not searched if you
+# override these settings, unless they are explicitly included.
+# The ``:`` is os.pathsep, which may vary on windows
```
Comment (pv) [inline site.cfg.example]: "..., which is ``;`` on Windows."

Comment (mattip) [reply site.cfg.example]: changing

**Comment by dimpase** [comment]
another seemingly undocumented feature/bug:
```
[blas]
libraries = foo, bar
``` 
is not the same as
```
[blas]
blas_libs = foo, bar
```
The latter leads to both l`ibfoo` and `libbar` used in linking (the intention), the former does not.
(see https://trac.sagemath.org/ticket/29051 for details; namely, on Archlinux the complete BLAS functionality is provided by two libs: (Open)BLAS and cblas).

**Comment by pv** [comment]
The `libraries` thing -> separate issue, thanks.

**Review by charris — APPROVED** [review]
I was just waiting for Pauli to sign off on this.

**Comment by charris** [comment]
Note that the `djbfft` section was from long ago when scipy supported different backends for the fft. See https://scipy-dev.scipy.narkive.com/s9YDkiZk/dropping-djbfft from 12 years ago. FFTW was another package that used to be supported. I don't think we can use either at this point so we might want to remove them. But that can be a different PR.

**Comment by charris** [comment]
Thanks Matti.

**Comment by dimpase** [comment]
> The `libraries` thing -> separate issue, thanks.

please see #15481
